File: sensores/smartv.py
import pika
import time
import json

# Configurações
RABBITMQ_HOST = 'localhost'
QUEUE_NAME = 'fila_smartv'
FILE_PATH = '../servidor/tvstate.json'

# Função para ler o estado da TV do arquivo JSON
def read_state_from_file():
    try:
        with open(FILE_PATH, 'r') as file:
            return json.load(file)
    except (FileNotFoundError, json.JSONDecodeError) as e:
        print(f"Erro ao ler o estado da TV: {e}")
        return None

# Função para publicar o estado da TV no RabbitMQ
def publish_state(channel):
    state = read_state_from_file()
    if state:
        message = json.dumps(state)
        channel.basic_publish(
            exchange='',
            routing_key=QUEUE_NAME,
            body=message,
            properties=pika.BasicProperties(delivery_mode=2)  # Torna a mensagem persistente
        )
        print(f"📤 Estado da TV publicado na fila: {message}")

# Função principal
def start_sensor():
    try:
        # Conecta ao RabbitMQ
        connection = pika.BlockingConnection(pika.ConnectionParameters(host=RABBITMQ_HOST))
        channel = connection.channel()

        # Declara a fila (se não existir, ela será criada)
        channel.queue_declare(queue=QUEUE_NAME, durable=True)

        print('Conexão com RabbitMQ estabelecida.')

        # Publica o estado da TV a cada 5 segundos
        while True:
            publish_state(channel)
            time.sleep(5)

    except Exception as e:
        print(f"Erro ao iniciar o sensor: {e}")
    finally:
        if 'connection' in locals() and connection.is_open:
            connection.close()

if __name__ == '__main__':
    print('TV iniciado, publicando estado da TV no RabbitMQ...')
    start_sensor()

File: sensores/smartv.js
const amqp = require('amqplib');
const fs = require('fs');

let printCounter = 0;
FILE_PATH = './servidor/tvstate.json';
RABBITMQ_HOST = 'localhost'
QUEUE_NAME = 'fila_smartv'

//Ler o estado da TV do arquivo JSON
function read_state_from_file() {
    try {
        const data = fs.readFileSync(FILE_PATH, 'utf8');
        return JSON.parse(data);
    } catch (error) {
        console.error("Erro ao ler o estado da TV:", error);
        return 
    }
}

async function connectRabbitMQ() {
    while (true) {
        try {
            const connection = await amqp.connect('amqp://localhost');
            const channel = await connection.createChannel();
            await channel.assertQueue('fila_smartv', { durable: true });
            console.log('Conexão com RabbitMQ estabelecida.');
            return channel;
        } catch (error) {
            console.log(`Erro ao conectar ao RabbitMQ: ${error}. Tentando novamente...`);
            await new Promise(resolve => setTimeout(resolve, 5000));
        }
    }
}


async function publishState(channel) {
    const state = read_state_from_file();
    await channel.sendToQueue('fila_smartv', Buffer.from(JSON.stringify(state)), { persistent: true });
    console.log(`📤 Estado da TV publicado (${printCounter}) na fila: ${JSON.stringify(state)}`);
    printCounter++;
}


function startPeriodicPublishing(channel) {
    setInterval(async () => {
        await publishState(channel);
    }, 5000);  
}

async function startSensor() {
    try {
        const channel = await connectRabbitMQ();
        startPeriodicPublishing(channel);
    } catch (error) {
        console.log(`Erro ao iniciar o sensor: ${error}`);
    }
}

startSensor().then(() => {
    console.log('TV iniciado, publicando estado da TV no RabbitMQ...');
});